### PR TITLE
SONAR-8716 clean-up UserSession

### PR DESCRIPTION
**Pull request by simonbrandhof**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/CurrentAction.java [hunk 1]
```diff
@@ -126,17 +130,20 @@ private static void writeGroups(JsonWriter json, Collection<String> groups) {
...
+  private void writeGlobalPermissions(JsonWriter json) {
     json.name("global").beginArray();
-    for (String permission : session.globalPermissions()) {
-      json.value(permission);
-    }
+
+    String defaultOrganizationUuid = defaultOrganizationProvider.get().getUuid();
+    GlobalPermissions.ALL.stream()
+      .filter(permission -> userSession.hasOrganizationPermission(defaultOrganizationUuid, permission))
+      .forEach(permission -> json.value(permission));
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/CurrentAction.java]: MINOR Replace this lambda with a method reference. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1612)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: server/sonar-server/src/test/java/org/sonar/server/tester/AbstractMockUserSession.java [hunk 2]
```diff
@@ -69,13 +69,16 @@ public T addComponentUuidPermission(String projectPermission, String projectUuid
 
   @Override
   public boolean hasComponentPermission(String permission, ComponentDto component) {
-    return hasComponentUuidPermission(permission, component.projectUuid());
+    return isRoot() || hasComponentUuidPermission(permission, component.projectUuid());
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/tester/AbstractMockUserSession.java]: unnecessary change as `hasComponentUuidPermission` also checks `isRoot()`

File: server/sonar-server/src/test/java/org/sonar/server/tester/AbstractMockUserSession.java [hunk 3]
```diff
@@ -84,7 +87,7 @@ private boolean hasProjectPermissionByUuid(String permission, String projectUuid
 
   @Override
   public boolean hasOrganizationPermission(String organizationUuid, String permission) {
-    return permissionsByOrganizationUuid.get(organizationUuid).contains(permission);
+    return isRoot() || permissionsByOrganizationUuid.get(organizationUuid).contains(permission);
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/tester/AbstractMockUserSession.java]: as discussed earlier, this may have to be dropped

File: server/sonar-server/src/main/resources/org/sonar/server/user/ws/current-example.json [hunk 4]
```diff
@@ -15,11 +15,8 @@
   ],
   "permissions": {
     "global": [
-      "admin",
```
Comment (sns-seb) [inline server/sonar-server/src/main/resources/org/sonar/server/user/ws/current-example.json]: why removing some permissions?
I guess the point of the example was also to show all available permissions (which I think is not a bad idea even if the WS is internal)

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/CurrentAction.java [hunk 5]
```diff
@@ -25,9 +25,11 @@
 import org.sonar.api.server.ws.Response;
 import org.sonar.api.server.ws.WebService.NewController;
 import org.sonar.api.utils.text.JsonWriter;
+import org.sonar.core.permission.GlobalPermissions;
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/CurrentAction.java]: commit title "SONAR-8716 api/users/current must return permissions on default organization" too long

File: sonar-db/src/test/java/org/sonar/db/component/ComponentDaoTest.java [hunk 6]
```diff
@@ -527,21 +526,23 @@ public void select_projects_from_view() {
...
+      .extracting(ComponentDto::uuid)
+      .containsOnly(provisionedProject.uuid(), projectUuid);
   }
 
   @Test
   public void select_provisioned() {
     OrganizationDto organization = db.organizations().insert();
     ComponentDto provisionedProject = db.components().insertComponent(newProjectDto(organization).setKey("provisioned.project").setName("Provisioned Project"));
-    ComponentDto provisionedView = db.components().insertView(organization, (dto) -> {});
+    ComponentDto provisionedView = db.components().insertView(organization, (dto) -> {
```
Comment (sns-seb) [inline sonar-db/src/test/java/org/sonar/db/component/ComponentDaoTest.java]: FYI, an override of `insertView` without a second argument exists
(it's legacy code, I know)

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/TaskAction.java [hunk 7]
```diff
@@ -120,7 +122,12 @@ public void handle(Request wsRequest, Response wsResponse) throws Exception {
 
   private void checkPermission(Optional<ComponentDto> component) {
     if (component.isPresent()) {
-      userSession.checkComponentPermission(SCAN_EXECUTION, component.get());
+      if (!userSession.hasOrganizationPermission(component.get().getOrganizationUuid(), GlobalPermissions.SYSTEM_ADMIN) &&
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/TaskAction.java]: put `component.get().getOrganizationUuid()` into a variable for readability, perf and to ensure the same default organization is checked (even though the implementation is cached, it's not in the contract of the interface)

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/TaskAction.java [hunk 8]
```diff
@@ -120,7 +122,12 @@ public void handle(Request wsRequest, Response wsResponse) throws Exception {
...
+        !userSession.hasOrganizationPermission(component.get().getOrganizationUuid(), GlobalPermissions.SCAN_EXECUTION) &&
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/TaskAction.java]: `SCAN_EXECUTION` is already statically imported (for the next line), let's use it! (and statically import  `SYSTEM_ADMIN` for consistency)

File: server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java [hunk 9]
```diff
@@ -113,6 +113,9 @@ public boolean isRoot() {
 
   @Override
   public boolean hasOrganizationPermission(String organizationUuid, String permission) {
+    if (isRoot()) {
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java]: as discussed earlier, this might have to be rolledback

File: server/sonar-server/src/test/java/org/sonar/server/user/ServerUserSessionTest.java [hunk 10]
```diff
@@ -203,6 +203,13 @@ public void hasOrganizationPermission_for_logged_in_user() {
   }
 
   @Test
+  public void hasOrganizationPermission_always_returns_true_for_roots() {
+    UserSession session = newUserSession(ROOT_USER_DTO);
+    assertThat(session.hasOrganizationPermission("an-uuid", GlobalPermissions.PROVISIONING)).isTrue();
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/user/ServerUserSessionTest.java]: when testing an "always" condition such as here, you might want to loop over all global permissions rather than just testing a few of them (it's possible thanks to `GlobalPermissions.ALL`)

File: server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java [hunk 11]
```diff
@@ -25,18 +25,31 @@
...
+
+  /**
+   * Login of the authenticated user. Returns {@code null}
+   * if {@link #isLoggedIn()} is {@code false}.
+   */
   @CheckForNull
   String getLogin();
 
+  /**
+   * NAme of the authenticated user. Returns {@code null}
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java]: typo `NAme` => `Name`

File: server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java [hunk 12]
```diff
@@ -58,47 +74,57 @@
...
+   * Returns {@code true} if the permission is granted to user on the component,
+   * otherwise {@code false}.
+   *
+   * If the component does not exist, then returns {@code false}.
    *
-   * @see org.sonar.api.web.UserRole for list of project permissions
+   * Always returns {@code true} if {@link #isRoot()} is {@code true}, even if
+   * component does not exist.
+   *
+   * If the permission is not granted, then the organization permission is _not_ checked.
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java]: you could use `<strong>not</strong>` instead of `_not_`. It's more how it's supposed to be done in Javadoc

File: server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java [hunk 13]
```diff
@@ -58,47 +74,57 @@
...
   /**
-   * Ensures that permission is granted to user, otherwise throws a {@link org.sonar.server.exceptions.ForbiddenException}.
-   * If the component doesn't exist and the user doesn't have the permission, throws
-   * a {@link org.sonar.server.exceptions.ForbiddenException}.
+   * Using {@link #hasComponentPermission(String, ComponentDto)} is recommended
+   * because it does not have to load project if the referenced component
+   * is not a project.
    */
-  UserSession checkComponentUuidPermission(String permission, String componentUuid);
+  boolean hasComponentUuidPermission(String permission, String componentUuid);
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java]: I wonder if this shouldn't just be deprecated

File: server/sonar-server/src/main/resources/org/sonar/server/user/ws/current-example.json (hunk 4 above)
Comment (simonbrandhof) [inline server/sonar-server/src/main/resources/org/sonar/server/user/ws/current-example.json]: else it does not test that filtering is applied. It probably deserves another test.

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/CurrentAction.java [hunk 14]
```diff
@@ -103,7 +107,7 @@ private void writeUserDetails(JsonWriter json, Optional<UserDto> optionalUser, C
 
     writeScmAccounts(json, optionalUser);
     writeGroups(json, groups);
-    writePermissions(json, userSession);
+    writePermissions(json);
   }
 
   private static void writeScmAccounts(JsonWriter json, Optional<UserDto> optionalUser) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/CurrentAction.java]: MINOR Specify a "UserDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [CurrentAction.java#L87](https://github.com/SonarSource/sonarqube/blob/49cc1aae7ba4134096dcec10e748cc14081bd570/server/sonar-server/src/main/java/org/sonar/server/user/ws/CurrentAction.java#L87): Specify a "UserDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)
1. ![MINOR][MINOR] [CurrentAction.java#L93](https://github.com/SonarSource/sonarqube/blob/49cc1aae7ba4134096dcec10e748cc14081bd570/server/sonar-server/src/main/java/org/sonar/server/user/ws/CurrentAction.java#L93): Specify a "UserDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'
